jcl/src/java.base/share/classes/com/ibm/tools/attach/target/IPC.java

```diff
@@ -365,49 +394,50 @@ public static void logMessage(String string1, int int1, String string2) {
 	 * @param string3 third string
 	 */
 	public static void logMessage(String string1, int int1, String string2, String string3) {
-		if (loggingEnabled ) {
+		if (isLoggingEnabled()) {
 			printLogMessage(string1+Integer.toString(int1)+string2+string3);
 		}
 	}
 
-	private static void printMessageWithHeader(String msg, PrintStream log) {
-		tracepoint(TRACEPOINT_STATUS_LOGGING, msg);
-		printLogMessageHeader(log);
-		log.println(msg);
-	}
-
 	/**
 	 * Print the information about a throwable, including the exact class,
 	 * message, and stack trace.
 	 * @param msg User supplied message
 	 * @param thrown throwable
 	 * @note nothing is printed if logging is disabled
 	 */
-	public synchronized static void logMessage(String msg, Throwable thrown) {
-		@SuppressWarnings("resource") /* this will be closed when the VM exits */
-		PrintStream log = getLogStream();
-		if (!Objects.isNull(log)) {
-			printMessageWithHeader(msg, log);
-			thrown.printStackTrace(log);
-			log.flush();
+	public static void logMessage(String msg, Throwable thrown) {
+		synchronized (accessorMutex) {
+			if (isLoggingEnabled()) {
+				printMessageWithHeader(msg, logStream);
+				thrown.printStackTrace(logStream);
+				logStream.flush();
+			}
 		}
 	}
 
 	/**
 	 * Print a message to the log file with time and thread information.
 	 * Also send the raw message to a tracepoint.
+	 * Call this only if isLoggingEnabled() has returned true.
 	 * @param msg message to print
 	 * @note no message is printed if logging is disabled
 	 */
-	private synchronized static void printLogMessage(final String msg) {
-		@SuppressWarnings("resource") /* this will be closed when the VM exits */
-		PrintStream log = getLogStream();
-		if (!Objects.isNull(log)) {
-			printMessageWithHeader(msg, log);
-			log.flush();
+	private static void printLogMessage(final String msg) {
+		synchronized (accessorMutex) {
+			if (!Objects.isNull(logStream)) {
+				printMessageWithHeader(msg, logStream);
+				logStream.flush();
+			}
 		}
 	}
 
+	private static void printMessageWithHeader(String msg, PrintStream log) {
```

Comment:
This method should have a comment that it must be called from inside a synchronized block.

Reply:
Good plan.